group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/assignor/StickyTaskAssignor.java

```diff
@@ -0,0 +1,401 @@
+            .min(Map.Entry.comparingByValue())
+            .map(Map.Entry::getKey);
+        return memberWithLeastLoad.map(memberId -> new Member(processWithLeastLoad.processId(), memberId)).orElse(null);
+    }
+
+    private boolean hasUnfulfilledQuota(final Member member) {
+        return processIdToState.get(member.processId).memberToTaskCounts().get(member.memberId) < tasksPerMember;
+    }
+
+    private void assignStandby(final Set<TaskId> standbyTasks, final int numStandbyReplicas) {
+        for (TaskId task : standbyTasks) {
+            for (int i = 0; i < numStandbyReplicas; i++) {
+                final Set<String> availableProcesses = findAllowedProcesses(task);
+                if (availableProcesses.isEmpty()) {
+                    log.warn("Unable to assign " + (numStandbyReplicas - i) +
+                        " of " + numStandbyReplicas + " standby tasks for task [" + task + "]. " +
+                        "There is not enough available capacity. You should " +
+                        "increase the number of threads and/or application instances " +
+                        "to maintain the requested number of standby replicas.");
+                    break;
+                }
+                Member standby = null;
+                // prev active task
+                Member prevMember = activeTaskToPrevMember.get(task);
+                if (prevMember != null && availableProcesses.contains(prevMember.processId) && isLoadBalanced(prevMember.processId)
+                    && taskPairs.hasNewPair(task, processIdToState.get(prevMember.processId).assignedTasks())) {
+                    standby = prevMember;
+                }
+                // prev standby tasks
+                if (standby == null) {
+                    final Set<Member> prevMembers = standbyTaskToPrevMember.get(task);
+                    if (prevMembers != null && !prevMembers.isEmpty()) {
+                        prevMembers.removeIf(member  -> !availableProcesses.contains(member.processId));
+                        prevMember = findMemberWithLeastLoad(prevMembers, task, true);
+                        if (prevMember != null && isLoadBalanced(prevMember.processId)) {
+                            standby = prevMember;
+                        }
+                    }
+                }
+                // others
+                if (standby == null) {
+                    standby = findMemberWithLeastLoad(task, availableProcesses);
+                }
+                processIdToState.get(standby.processId).addTask(standby.memberId, task, false);
+                updateHelpers(standby, task, false);
+            }
+        }
+    }
+
+    private boolean isLoadBalanced(final String processId) {
+        final ProcessState process = processIdToState.get(processId);
+        return process.hasCapacity() || isLeastLoadedProcess(process.load());
+    }
+
+    private boolean isLeastLoadedProcess(final double load) {
```

Comment:
Another more general question - when it comes to calculating load we will consider stateful vs. stateless ?

Reply:
Very good question @bbejeck 
I think currently, we don't consider task type ([here](https://github.com/apache/kafka/blob/a8f6fc9cc4cf1db27f90145296d9d597995081ba/streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java#L337-L341)). 
But that would also be a good point for further discussion.